### PR TITLE
[5.1] SEF: Implementing trailing slash behavior

### PR DESCRIPTION
**Pull request by Hackwar**
### Summary of Changes
Joomla has been improving its SEO performance constantly and one issue which is still open is the behavior of trailing slashes in URLs. Right now, Joomla has largely been following a no-trailing-slashes policy, however we never enforced this. At the same time, some SEO people wanted to have trailing slashes on all URLs. This PR introduces this feature.

This introduces 2 new settings in the SEF system plugin. The first setting lets you decide if you want to have a trailing slash, no trailing slash or no special behavior (which is the b/c setting). Depending on the setting, it either adds or removes all trailing slashes.

The second setting is available when one of the former options is set and redirects a GET request with missing or trailing slash to the "correct" URL with a 301 depending on the other option.

This PR depends on #42692.

I'd like to thank [djumla GmbH](https://www.djumla.de) for sponsoring this feature.

### Testing Instructions
1. Apply the PR
2. Check that no URLs changed
3. Edit libraries/src/Router/SiteRouter.php and change line 466 to `$uri->setPath($uri->getPath() . '/' . $tmp . '/');`
4. See that all URLs now contain a trailing slash.
5. Go to the SEF system plugin in the backend and set the setting to remove the trailing slash.
6. See that all URLs now don't contain a trailing slash.
7. Revert the change from step 3.
8. Change the setting in the SEF plugin to always add a trailing slash.
9. See that all URLs now contain a traling slash again.
10. Set the second option to enforce the behavior with a 301 redirect.
11. Go to the frontend again and either add or remove the trailing slash, depending on your setting, in your browser bar.
12. See that Joomla redirected you to the same page with the correct URL again.

### Link to documentations
Please select:
- [X] Documentation link for docs.joomla.org: https://docs.joomla.org/Search_Engine_Friendly_URLs
- [ ] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [ ] No documentation changes for manual.joomla.org needed

### CODE REVIEW

**Comment by brianteeman** [comment]
1. Instead of adding another option which a large % of users will never use why not simply do this in the htaccess.
2. the description is confusing "Remove unneeded index.php from URLs" because it sounds the same as the existing option which removes index.php
3. If accepted dont you need to update the installation and the configuration.php.dist
4. What happened to the requirement for all new "features" to be accompanied with documentation

**Comment by Hackwar** [comment]
To 1.: it is a difference if we output wrong URLs and then redirect in the htaccess or do it correct right away. Yes, the redirect could be done in the htaccess, but that requires error prone actions from the user, while this here is the simpler solution.

To 2.: no, since these options are not stored in the configuration.php.

To 3.: the PR is not merged yet, is it. 😉 Documentation is on its way and will be combined with the other 2 router related PRs.

File: administrator/language/en-GB/plg_system_sef.ini [hunk 1]
```diff
@@ -5,5 +5,11 @@
 
 PLG_SEF_DOMAIN_DESCRIPTION="If your site can be accessed through more than one domain enter the preferred (sometimes referred to as canonical) domain here. <br><strong>Note:</strong> https://example.com and https://www.example.com are different domains."
 PLG_SEF_DOMAIN_LABEL="Site Domain"
+PLG_SEF_TRAILINGSLASH_DESCRIPTION="Force Joomla to only create URLs with or without trailing slash. This is only applied when 'Add suffix to URL' is disabled."
```
Comment (brianteeman) [inline administrator/language/en-GB/plg_system_sef.ini]: please add that this is in the global config as you have done in the other PR

**Comment by brianteeman** [comment]
I have tested this item :red_circle: unsuccessfully on 131eafeba08719bc4b6b1899004931399d554975<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42702">issues.joomla.org/tracker/joomla-cms/42702</a>.</sub>

**Comment by brianteeman** [comment]
Tested with both of the following
image
image

In both cases with the sample joomla data I observed no trailing slash in the url
image
image

**Comment by Hackwar** [comment]
Did you read that this PR depends on #42692 ?

**Comment by brianteeman** [comment]
no I didnt see that

**Comment by brianteeman** [comment]
I have not tested this item.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42702">issues.joomla.org/tracker/joomla-cms/42702</a>.</sub>

File: plugins/system/sef/src/Extension/Sef.php [hunk 2]
```diff
@@ -23,8 +27,65 @@
...
+     * After initialise router.
+     *
+     * @return  void
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    public function onAfterInitialiseRouter(AfterInitialiseRouterEvent $event)
+    {
+        if (
+            is_a($event->getRouter(), SiteRouter::class)
```
Comment (SniperSister) [inline plugins/system/sef/src/Extension/Sef.php]: Invert the condition for an early return, reduces nesting level

File: plugins/system/sef/src/Extension/Sef.php [hunk 3]
```diff
@@ -188,6 +249,74 @@ function ($match) use ($base, $protocols) {
...
+     * @param   Uri     &$uri     Uri object.
+     *
+     * @return  void
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    public function enforceTrailingSlash(&$router, &$uri)
+    {
+        // We only want to redirect on GET requests
+        if ($_SERVER['REQUEST_METHOD'] != 'GET') {
```
Comment (SniperSister) [inline plugins/system/sef/src/Extension/Sef.php]: `$this->app->getInput()->getMethod()` should be the framework equivalent

File: plugins/system/sef/src/Extension/Sef.php [hunk 4]
```diff
@@ -188,6 +249,74 @@ function ($match) use ($base, $protocols) {
...
+            return;
+        }
+
+        $origUri = Uri::getInstance();
```
Comment (SniperSister) [inline plugins/system/sef/src/Extension/Sef.php]: $originalUri - 4 more chars don't hurt ;)

File: administrator/language/en-GB/plg_system_sef.ini (hunk 1 above)
Comment (SniperSister) [reply administrator/language/en-GB/plg_system_sef.ini]: While talking about potential fixes for the issue, we also discussed where to place the new parameter. The general idea was to move these more "advanced" SEF settings to the plugin configuration, just like the already existing SEF domain parameter. Consider it like an "expert" section for SEF settings.

Comment (brianteeman) [reply administrator/language/en-GB/plg_system_sef.ini]: why is it even an option

Comment (SniperSister) [reply administrator/language/en-GB/plg_system_sef.ini]: You mean why not just enable forcing by default, once a behavior (append or remove slash) has been selected?

Comment (brianteeman) [reply administrator/language/en-GB/plg_system_sef.ini]: yes. If its correct to make this change then make it globaly. if the answer is "it depends" then make it an extension

Comment (Hackwar) [reply administrator/language/en-GB/plg_system_sef.ini]: Searching for `seo trailing slash` for me results in way to many differing opinions to force this in one direction. I personally would not add a trailing slash, but this is not about my preference.

Comment (Hackwar) [reply administrator/language/en-GB/plg_system_sef.ini]: Besides the point that I've been working on removing the trailing slash in the last 10 years wherever it came up in the core.

Comment (brianteeman) [reply administrator/language/en-GB/plg_system_sef.ini]: all the more reason to make this an extension and not to pollute the core with more code to maintain and more options to confuse users

Comment (SniperSister) [reply administrator/language/en-GB/plg_system_sef.ini]: We are mixing two things here:

* Question a: should Joomla allow users to force-append or force-remove trailing slashes from URLs that have been generated by the core
* Question b: should Joomla redirect incoming calls to match a selected "trailing-slash" pattern in option a

Question A is a definite "yes" for me! It should be possible to enforce a consistent behavior in the generation and I personally would very strongly vote to enforce one (couldn't care less which one) for **new** installations. It however has to been an option as an global enforcement that is applied automatically on update might be a b/c break for existing sites.

Question B: personally I don't need this to be an option. If I have made a decision to enforce a consistent generation pattern (see question A) then I think it should be enforced for incoming calls too. So that option is up to discussion for me.

Comment (jokorntheuer) [reply administrator/language/en-GB/plg_system_sef.ini]: Question a: yes
Question b: yes, because if we set this to no, exisiting external backlinks from other websites will point Google to URL-versions we do not want in the index. this feature is only consistens if a & b is a yes tbh.

**Comment by SniperSister** [comment]
I have tested this item :white_check_mark: successfully on d49ccdf9d47c3cb241e20738e5ed42825ff04286<br /><br />Work as described!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42702">issues.joomla.org/tracker/joomla-cms/42702</a>.</sub>

**Comment by viocassel** [comment]
I have tested this item :white_check_mark: successfully on d49ccdf9d47c3cb241e20738e5ed42825ff04286<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42702">issues.joomla.org/tracker/joomla-cms/42702</a>.</sub>

**Comment by richard67** [comment]
Setting RTC as it has 2 good tests, but setting also the RMDG (release manager decision queue) label because this PR is subject of discussion among maintainers.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42702">issues.joomla.org/tracker/joomla-cms/42702</a>.</sub>

File: plugins/system/sef/sef.xml [hunk 5]
```diff
@@ -30,6 +30,32 @@
...
+				<field
+					name="trailingslash_redirect"
+					type="radio"
+					label="PLG_SEF_TRAILINGSLASH_REDIRECT_LABEL"
+					layout="joomla.form.field.radio.switcher"
+					default="0"
+					filter="boolean"
+					showon="trailingslash:1,2"
+				>
+					<option value="0">JNO</option>
```
Comment (Quy) [inline plugins/system/sef/sef.xml]:
```suggestion
					>
					<option value="0">PLG_SEF_TRAILINGSLASH_OPTION_NONE</option>
					<option value="1">PLG_SEF_TRAILINGSLASH_OPTION_NO_SLASH</option>
					<option value="2">PLG_SEF_TRAILINGSLASH_OPTION_SLASH</option>
				</field>

				<field
					name="trailingslash_redirect"
					type="radio"
					label="PLG_SEF_TRAILINGSLASH_REDIRECT_LABEL"
					layout="joomla.form.field.radio.switcher"
					default="0"
					filter="boolean"
					showon="trailingslash:1,2"
					>
					<option value="0">JNO</option>
```

**Comment by Hackwar** [comment]
In the CMS Maintainer meeting we talked about this PR and decided to join the two options into one. So if you decide for one behavior, it automatically also means when the URL is not as expected, it does a redirect to the right one. I made the necessary changes. @SniperSister and @viocassel would you be able to test this again?

**Comment by richard67** [comment]
Back to pending<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42702">issues.joomla.org/tracker/joomla-cms/42702</a>.</sub>

File: administrator/language/en-GB/plg_system_sef.ini [hunk 6]
```diff
@@ -5,5 +5,10 @@
 
 PLG_SEF_DOMAIN_DESCRIPTION="If your site can be accessed through more than one domain enter the preferred (sometimes referred to as canonical) domain here. <br><strong>Note:</strong> https://example.com and https://www.example.com are different domains."
 PLG_SEF_DOMAIN_LABEL="Site Domain"
+PLG_SEF_TRAILINGSLASH_DESCRIPTION="Force Joomla to only use URLs with or without trailing slash. When set, this will force the right URL with redirects and is only applied when 'Add suffix to URL' is disabled."
+PLG_SEF_TRAILINGSLASH_LABEL="Trailing slash for URLs"
```
Comment (Quy) [inline administrator/language/en-GB/plg_system_sef.ini:9]:
```suggestion
PLG_SEF_TRAILINGSLASH_LABEL="Trailing Slash for URLs"
```

**Comment by viocassel** [comment]
I have tested this item :white_check_mark: successfully on baac8cd0bf3236a9b332a38e91ed433800e914ec<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42702">issues.joomla.org/tracker/joomla-cms/42702</a>.</sub>

**Comment by SniperSister** [comment]
I have tested this item :white_check_mark: successfully on baac8cd0bf3236a9b332a38e91ed433800e914ec<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42702">issues.joomla.org/tracker/joomla-cms/42702</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42702">issues.joomla.org/tracker/joomla-cms/42702</a>.</sub>

**Comment by richard67** [comment]
Back to pending as the PR has received a change. @SniperSister @viocassel Could you test again? Thanks in advance.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42702">issues.joomla.org/tracker/joomla-cms/42702</a>.</sub>

**Comment by SniperSister** [comment]
I have tested this item :white_check_mark: successfully on a9f1b3c6a391840860ba131d12a9ac41e35dffab<br /><br />Works as expected and fixes the issue caught by Martina<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42702">issues.joomla.org/tracker/joomla-cms/42702</a>.</sub>

**Comment by viocassel** [comment]
I have tested this item :white_check_mark: successfully on a9f1b3c6a391840860ba131d12a9ac41e35dffab<br /><br />👌<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42702">issues.joomla.org/tracker/joomla-cms/42702</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42702">issues.joomla.org/tracker/joomla-cms/42702</a>.</sub>

**Comment by brianteeman** [comment]
This should be retested as there has been a change

**Comment by Hackwar** [comment]
After some discussions, we decided to switch this to use 301 redirects instead of 303s.

**Comment by richard67** [comment]
Back to pending as there have been made changes.
@SniperSister @viocassel Could you test again? Thanks in advance.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42702">issues.joomla.org/tracker/joomla-cms/42702</a>.</sub>

**Comment by SniperSister** [comment]
I have tested this item :white_check_mark: successfully on 74383a3970a892403e2d51808138287f3b6166c7<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42702">issues.joomla.org/tracker/joomla-cms/42702</a>.</sub>

**Comment by viocassel** [comment]
I have tested this item :white_check_mark: successfully on 74383a3970a892403e2d51808138287f3b6166c7<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42702">issues.joomla.org/tracker/joomla-cms/42702</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42702">issues.joomla.org/tracker/joomla-cms/42702</a>.</sub>

**Comment by Hackwar** [comment]
Link to documentation has been added.

**Comment by LadySolveig** [comment]
Thank you @Hackwar and for testing and support @SniperSister @viocassel @richard67

**Comment by Hackwar** [comment]
Thank you!

**Comment by pAnd0rASBG** [comment]
Thank you very much for this @Hackwar ! I'm glad it didn't get dismissed.

my 2c on the importance:
according to Google's Martin Splitt, who I had a conversation with on this topic late last year, this still counts: https://developers.google.com/search/blog/2010/04/to-slash-or-not-to-slash?hl=en

tl;dr:
Trailing slashes do matter to Google. Traditionally that comes from url w/ trailing slash being seen as directory and url w/o trailing slash as file. Google respects same urls with and without trailing slashes to have different content, even. 
On the other hand, if e.g. https://mysite.com/blog and https://mysite.com/blog/ show the same content, Google may punish you for duplicate content and even if that's not the case, your "Link Juice" will be split 50/50, so instead of getting one better ranked entry in the SERPs, you get 2 worse.

So even if a large % of users will not actively use it, if there is a default behaviour for either with or without trailing slash, still all users will profit from the SEO impact.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42702">issues.joomla.org/tracker/joomla-cms/42702</a>.</sub>

**Comment by Hackwar** [comment]
You're welcome. ;-)

**Comment by rfmjoe** [comment]
Thank you Hackwar for bringing this finally into the joomla core! :-)

**Comment by brbrbr** [comment]
Adding slashes is not working. onAfterInitialise checks on 2 should be 1

**Comment by brbrbr** [comment]
Are you sure the slashes are working in the attachBuildRule part? Appears to be a bit of number bonanza on the trailingslash param: 1 and 2 in onAfterInitialise 0 and 1 in enforceTrailingSlash

**Comment by SniperSister** [comment]
@brbrbr see https://github.com/joomla/joomla-cms/pull/43292
